google/cloud/forseti/scanner/audit/external_project_access_rules_engine.py

```diff
@@ -0,0 +1,286 @@
+
+        Args:
+            global_configs (dict): Inventory configurations.
+        """
+        self.rule_book = (
+            ExternalProjectAccessRuleBook(
+                global_configs,
+                self._load_rule_definitions()))
+
+    # TODO: The naming is confusing and needs to be fixed in all scanners.
+    def find_policy_violations(self, user_email,
+                               project_ancestry,
+                               force_rebuild=False):
+        """Determine whether project ancestry violates rules.
+
+        Args:
+            user_email (str): The user's e-mail
+            project_ancestry (list): List of ancestries which turn
+                                     out to a list of resources.
+            force_rebuild (bool): Force the rebuild of the rule book
+
+        Returns:
+             generator: A generator of rule violations.
+        """
+
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+
+        violations = itertools.chain(
+            self.rule_book.find_policy_violations(
+                user_email, project_ancestry)
+        )
+
+        return violations
+
+
+class ExternalProjectAccessRuleBook(bre.BaseRuleBook):
+    """The RuleBook for ExternalProjectAccess resources."""
+
+    # Class variable for matching the ancestor during rule validation
+    ancestor_pattern = re.compile(r'^organizations/\d+$|^folders/\d+$')
+
+    def __init__(self, global_configs=None, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitons
+            global_configs (dict): Inventory configurations.
+        """
+        super(ExternalProjectAccessRuleBook, self).__init__()
+        self.resource_rules_map = dict()
+
+        # Create a default rule for the current organization as
+        # defined in the server inventory configuration.
+        root_resource_id = global_configs.get_root_resource_id()
+
+        default_rule = dict(name='Default',
+                            allowed_ancestors=[root_resource_id])
+
+        if not rule_defs:
+            rule_defs = dict(rules=[])
+
+        rule_defs['rules'].insert(0, default_rule)
+        self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book.
+
+        Args:
+            rule_defs (dict): rule definitions dictionary
+        """
+
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        ancestors = ExternalProjectAccessRuleBook.process_rule(rule_def,
```

Comment:
This should not be another classmethod.  You should just go with instance method:
```
    ancestors = self.process_rule(rule_def, rule_index)
```
And remove the @classmethod decorator, and go like with `self`:
```
def process_rule(self, rule_def, rule_index):
```